File: scrapper/scrapper.mjs

import { JSDOM } from "jsdom";
import fetch from 'node-fetch'
import fs from "fs"
import path from 'path'

// Obtener la fecha actual
const now = new Date();

// Restarle 90 días a la fecha actual
const limitDate = new Date(now.getTime() - (90 * 24 * 60 * 60 * 1000));

const urlOriginImage = "https://apod.nasa.gov/apod/"

// Iterar sobre cada fecha en el rango de fechas desde la fecha límite hasta la fecha actual
for (let i = 0, date = limitDate; i < 90; i++, date.setDate(date.getDate() + 1)) {
  const fecha = new Date(date);
  const fechaFormateada = fecha.toISOString().slice(0, 10);
  // imprime la fecha formateada '2023-02-08'
  // console.log(fechaFormateada); 

  // Construir la URL de la página de APOD correspondiente a cada fecha
  const url = `https://apod.nasa.gov/apod/ap${date
    .toISOString()
    .slice(2, 10)
    .replace(/-/g, '')}.html`;
    

  // reemplaza con la ruta a tu directorio de salida
  const currentDir = process.cwd(); 

  JSDOM.fromURL(url).then(dom => {
    const links = dom.window.document.querySelectorAll('a[href]');
    links.forEach(link => {
      const href = link.getAttribute('href');
      if (href.match(/image\/.*\.(jpeg|jpg|png)$/i)) {
        fetch(`${urlOriginImage}${href}`)
          .then(res => {
            const filename = `Day ${fechaFormateada}.jpg`;
            const dest = fs.createWriteStream(path.join(currentDir, filename));
            res.body.pipe(dest);
          })
          .catch(err => console.error(err));
      }
    });
  }).catch(err => console.error(err));
}

File: scrapper/scrapper.py
import requests
from datetime import datetime, timedelta
from bs4 import BeautifulSoup
import os
import re

# Obtener la fecha actual
now = datetime.now()

# Restarle 90 días a la fecha actual
limit_date = now - timedelta(days=90)

url_origin_image = "https://apod.nasa.gov/apod/"

# Iterar sobre cada fecha en el rango de fechas desde la fecha límite hasta la fecha actual
for i in range(90):
    date = limit_date + timedelta(days=i)
    formatted_date = date.strftime("%Y%m%d")
    # Agregar los guiones
    formatted_date_organized = '-'.join([formatted_date[:4], formatted_date[4:6], formatted_date[6:]])
    # Imprimir la fecha formateada '2023-02-08'
    # print(formatted_data_organized)

    # Construir la URL de la página de APOD correspondiente a cada fecha
    url = f"https://apod.nasa.gov/apod/ap{formatted_date[2:]}.html"

    response = requests.get(url)
    soup = BeautifulSoup(response.content, 'html.parser')

    for link in soup.find_all('a'):
        href = link.get('href')
        if href and re.match(r'image/.*\.(jpeg|jpg|png)$', href):
            response = requests.get(f"{url_origin_image}{href}")
            filename = f"Day {formatted_date_organized}.jpg"
            with open(os.path.join(os.getcwd(), filename), 'wb') as f:
                f.write(response.content)